### DISCUSSION TITLE
beforeNavigate fires when you click a link to the page you're on

### DISCUSSION THREAD

**Discussion by Blue-Nomad** [sveltejs/kit#12861] [Q&A] [answered]
The docs say "A navigation interceptor that triggers before we navigate to a **new** URL" but that doesn't seem to be the case.

I'm using beforeNavigate to clear IntersectionObserver objects like:

```
beforeNavigate(() =>
{
  observer.disconnect()
})
```

And when I'm on the home page and I click a link to the home page, that fires and kills the observer so it doesn't work on the page.

I could do:

```
beforeNavigate(({ from, to }) =>
{
  if (from.url.pathname === to.url.pathname) return
  observer.disconnect()
))
```

But that feels like it shouldn't be necessary.

Is this a bug with beforeNavigate or am I doing this in a weird way?

**Comment by brunnerh** [comment]
Can't you use [regular Svelte hooks](https://svelte.dev/docs/svelte/lifecycle-hooks) like `onMount`/`onDestroy`?
(`onDestroy` runs on the server as well, so often the cleanup function _returned_ by the callback in `onMount` is better.)

Also, how is the observer created? If it's `onMount`, that is also the perfect place to clean it up, if it's an [action](https://svelte.dev/docs/svelte/svelte-action), those also have a `destroy` function as part of the return object for cleanup.

**Reply by Blue-Nomad** [reply]
I found that onMount doesn't always run if 1 component is on 2 pages. I'm showing projects like /projects/project-1 and /projects/project-2 and both have a testimonial on them and when flicking back and forth between them, onMount only ran for the first project.

I need a way to force the setup and cleanup processes between pages that can have the same components on them. I was creating the observer inside afterNavigate for this reason.

Otherwise this works well, if going from a page with a testimonial to a page without one.

**Reply by brunnerh** [reply]
Can you create a _minimal_, complete example of what you are trying to do here?

**Reply by Blue-Nomad** [reply]
Sure, I've created https://github.com/Blue-Nomad/SvelteKit-Issue-Demo-1/ that I think shows this well.

When navigating from the Home page to a Project page and vice versa, all 4 hooks run, but when navigating between the Project pages the Mount hooks don't run.

So I'm looking for hooks that run every time a page is navigated to when 1 component could be on every page. beforeNavigate & afterNavigate seemed perfect for that until I realised they also run when navigating to the page you're already on, which in my case runs the cleanup functions and kills the functionality of the page.

**Reply by brunnerh** [reply]
This is still missing the core piece: The `IntersectionObeserver` and how it is used.

I don't really understand the problem — if the *route* is the same, can't you keep the observer?
Isn't this what you want, initialize the observer and only clean it up if you navigate to a different route?

How does the other component factor into this, and if it's relevant to the observer, why doesn't the observer live there?

**Reply by Blue-Nomad** [reply]
Thanks for the reply. I didn't realise onMount was tied to routes like that.

Now when a component is only on 1 route, I'm using onMount and the return function to observe and unobserve elements, and that's working well.

When a component is on multiple pages within 1 route and onMount won't run when moving between them, I'm using:

```
onMount(() =>
{
    observer = new IntersectionObserver(...)
})

beforeNavigate(() =>
{
    observer.unobserve(element)
})

afterNavigate(() =>
{
    observer.observe(element)
})
```

**Reply by brunnerh** [reply, marked as answer]
Why not use an action on the element the observer is tied to?

**Reply by Blue-Nomad** [reply]
That makes sense, then it's tied to the element and can observe/unobserve based on its existence. Thanks for the help.
